src/tools/compiletest/src/header/tests.rs

```diff
@@ -667,3 +667,24 @@ fn test_non_rs_unknown_directive_not_checked() {
     );
     assert!(!poisoned);
 }
+
+#[test]
+fn test_trailing_directive() {
+    let mut poisoned = false;
+    run_path(&mut poisoned, Path::new("a.rs"), b"//@ only-x86 only-arm");
+    assert!(poisoned);
+}
+
+#[test]
+fn test_trailing_directive_with_comment() {
+    let mut poisoned = false;
+    run_path(&mut poisoned, Path::new("a.rs"), b"//@ only-x86   only-arm with comment");
+    assert!(poisoned);
+}
+
+#[test]
+fn test_not_trailing_directive() {
+    let mut poisoned = false;
+    run_path(&mut poisoned, Path::new("a.rs"), b"//@ revisions: incremental");
```

Comment:
would this reject `//@ revisions incremental`? I want to say that the `:` was not enforced in compiletest directive grammar when parsing directives...

Reply:
No it would not be rejected, because it's currently used in `tests/ui/invalid-compile-flags/fuel.rs`:
https://github.com/rust-lang/rust/blob/b3bd7058c139e71bae0862ef8f8ac936208873e9/tests/ui/invalid-compile-flags/fuel.rs#L1

I also don't see a reason to reject it, since the `:` is a delimiter.

Reply:
Sorry let me clarify what I mean. I believe compiletest accepts both `//@ revisions incremental` and `//@ revisions: incremental`, would the logic in this PR (incorrectly) reject `//@ revisions incremental`?

Reply:
Yeah, unfortunately,

```
//@ revisions incremental
fn main() {}
```

would get rejected

```
error: detected trailing compiletest test directive `incremental`
```

whereas

```
//@ revisions: incremental
fn main() {}
```

the thing here is that compiletest accepts *both* forms (unfortunate directive grammar strikes yet again)

Reply:
No, no, it's me. I read the question the wrong way round.

> I believe compiletest accepts both `//@ revisions incremental` and `//@ revisions: incremental`

I just checked, and it doesn't seems to be the case. There is no revisions (also no errors).

> would the logic in this PR (incorrectly) reject `//@ revisions incremental`?

Yes, it would be rejected by this PR; but I would also argue that it's not incorrect since the syntax is a nop and should probably be an error.

Reply:
> I just checked, and it doesn't seems to be the case. There is no revisions (also no errors).

**What**

Reply:
Ah darn, I think it's because of a combination of things:

1. `revisions` is a known compiletest directive
2. when we parse name-value directives, the directive name is *required* to have a trailing `colon` in order for the parse to succeed: https://github.com/rust-lang/rust/blob/b14d8b2ef20c64c1002e2c6c724025c3d0846b91/src/tools/compiletest/src/header.rs#L1137-L1146

**However**, the way compiletest directive is setup right now means that if known directive check allows a directive but the corresponding name-value directive parsing logic fails, we **silently** accept the directive but **it has no effect**.

This is awful LOL

Reply:
Given that the colon is actually mandatory for directives like `revisions`, I think it's okay to reject things like `//@ only-arch known-directive` because the trailing known-directive-looking-thing is very unlikely to be a directive commment.

Reply:
I filed an issue https://github.com/rust-lang/rust/issues/123760 to make sure we know that this is a bug, but it's out of the scope for this PR. I'm happy to accept this PR as is.